src/compiler/types.rs

```diff
@@ -2714,31 +2816,57 @@ pub struct DiagnosticMessage {
 #[derive(Clone, Debug)]
 pub struct DiagnosticMessageChain {
     pub message_text: String,
+    pub code: u32,
     pub next: Option<Vec<DiagnosticMessageChain>>,
 }
 
-#[derive(Debug)]
+impl DiagnosticMessageChain {
+    pub fn new(message_text: String, code: u32, next: Option<Vec<DiagnosticMessageChain>>) -> Self {
+        Self {
+            message_text,
+            code,
+            next,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
 pub enum Diagnostic {
     DiagnosticWithLocation(DiagnosticWithLocation),
     DiagnosticWithDetachedLocation(DiagnosticWithDetachedLocation),
 }
 
-pub trait DiagnosticInterface: DiagnosticRelatedInformationInterface {}
+pub trait DiagnosticInterface: DiagnosticRelatedInformationInterface {
+    fn maybe_related_information(&self) -> Option<&[DiagnosticRelatedInformation]>;
+}
 
 #[derive(Clone, Debug)]
 pub struct BaseDiagnostic {
     _diagnostic_related_information: BaseDiagnosticRelatedInformation,
+    related_information: Option<Vec<DiagnosticRelatedInformation>>,
 }
 
 impl BaseDiagnostic {
-    pub fn new(diagnostic_related_information: BaseDiagnosticRelatedInformation) -> Self {
+    pub fn new(
+        diagnostic_related_information: BaseDiagnosticRelatedInformation,
+        related_information: Option<Vec<DiagnosticRelatedInformation>>,
+    ) -> Self {
         Self {
             _diagnostic_related_information: diagnostic_related_information,
+            related_information,
         }
     }
 }
 
 impl DiagnosticRelatedInformationInterface for BaseDiagnostic {
+    fn maybe_as_diagnostic(&self) -> Option<&Diagnostic> {
+        panic!("Shouldn't call maybe_as_diagnostic() on BaseDiagnostic")
```

Comment:
This is kind of messy, in the Typescript version `DiagnosticRelatedInformation` is the base interface for `Diagnostic`, before I was just exposing a `DiagnosticRelatedInformationInterface` that `Diagnostic` implemented but now it looks like there are concrete `DiagnosticRelatedInformation`'s getting stored/passed around so to avoid `dyn DiagnosticRelatedInformationInterface`'s (that then I wouldn't be able to extract the sub-variant of if needed) I went ahead and made `DiagnosticRelatedInformation` a concrete "parent" enum of `Diagnostic`. But it seems like most places are just using `Diagnostic`'s so I don't really want to always wrap in/pass around the parent `DiagnosticRelatedInformation`. So then what I'm doing now is making things generic on `<TDiagnosticRelatedInformation: DiagnosticRelatedInformationInterface>` and then if I need to do something specific to `Diagnostic`'s rather than being able to `match` on the parent `DiagnosticRelatedInformation` I exposed a `maybe_as_diagnostic()` method on `DiagnosticRelatedInformationInterface` that returns `Some(diagnostic)` for `Diagnostic` and `None` for the other variant of `DiagnosticRelatedInformation` (`BaseDiagnosticRelatedInformation`). But the intent is only to call `maybe_as_diagnostic()` on these "top-level" implementors of `DiagnosticRelatedInformationInterface` so I'm panicking if it gets called on anything besides those. I guess another alternative would be to make diagnostics use the same pattern as `Node`/`Type` and always wrap them and have a `.diagnostic_wrapper()` method that returns the parent `DiagnosticRelatedInformation` (from which I could then determine if it's a `Diagnostic` or not by `match`'ing)